File: platform/vapi/lib/drift.ts
/**
 * Merge-based drift detection (doc 08 §8.1).
 *
 * A naive `local` vs `remote` diff is permanently red: Vapi materialises every server
 * default and adds new ones over time. Instead we compare `remote` against
 * `deepMerge(remote, local)` — so drift is non-empty **iff a key we actually declare has a
 * different value remotely**. Keys the server adds and we don't declare are invisible.
 */

export type Json = null | boolean | number | string | Json[] | { [k: string]: Json };

const VOLATILE = new Set([
  'id',
  'orgId',
  'createdAt',
  'updatedAt',
  'isServerUrlSecretSet',
  'credentialId', // env-injected, instance-specific — masked, not compared
]);

/** Paths whose array order is server-assigned and meaningless. */
const SORT_ARRAYS_AT = new Set(['toolIds', 'structuredOutputIds', 'serverMessages']);

/**
 * Any drift here is a hard failure, never a warning: these are the fields that carry
 * compliance or routing meaning (doc 08 §8.1).
 */
export const FORBIDDEN_DRIFT = [
  'firstMessage',
  'serverMessages',
  'server.url',
  'model.messages.0.content',
  'compliancePlan.hipaaEnabled',
  'compliancePlan.pciEnabled',
  'artifactPlan.transcriptPlan.enabled',
];

export function isObject(v: unknown): v is Record<string, Json> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

/** `local` overlays `remote`. Arrays are replaced wholesale, never merged element-wise. */
export function deepMerge(remote: Json, local: Json): Json {
  if (!isObject(remote) || !isObject(local)) return local;
  const out: Record<string, Json> = { ...remote };
  for (const [k, v] of Object.entries(local)) {
    const prev = out[k];
    out[k] = isObject(prev) && isObject(v) ? deepMerge(prev, v) : v;
  }
  return out;
}

/** Strips volatile keys, sorts unordered arrays, and canonicalises for comparison. */
export function normalise(value: Json, key = ''): Json {
  if (Array.isArray(value)) {
    const items = value.map((v) => normalise(v));
    if (SORT_ARRAYS_AT.has(key)) {
      return [...items].sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
    }
    return items;
  }
  if (isObject(value)) {
    const out: Record<string, Json> = {};
    for (const k of Object.keys(value).sort()) {
      if (VOLATILE.has(k)) continue;
      const v = value[k];
      if (v === undefined || v === null) continue; // null ≡ absent
      out[k] = normalise(v, k);
    }
    return out;
  }
  if (typeof value === 'string') return value.trimEnd();
  return value;
}

export interface DriftEntry {
  path: string;
  remote: Json | undefined;
  desired: Json | undefined;
  forbidden: boolean;
}

function isForbidden(path: string): boolean {
  return FORBIDDEN_DRIFT.some((f) => path === f || path.startsWith(`${f}.`));
}

/** Reports only paths present in `desired` — i.e. paths we declare. */
export function diff(remote: Json, desired: Json, base = ''): DriftEntry[] {
  const out: DriftEntry[] = [];

  if (isObject(remote) && isObject(desired)) {
    for (const k of Object.keys(desired)) {
      out.push(...diff(remote[k] ?? null, desired[k] ?? null, base ? `${base}.${k}` : k));
    }
    return out;
  }

  if (JSON.stringify(remote) !== JSON.stringify(desired)) {
    out.push({
      path: base,
      remote: remote === null ? undefined : remote,
      desired: desired === null ? undefined : desired,
      forbidden: isForbidden(base),
    });
  }
  return out;
}

export function computeDrift(remoteRaw: Json, localRaw: Json): DriftEntry[] {
  const remote = normalise(remoteRaw);
  const desired = normalise(deepMerge(remoteRaw, localRaw));
  return diff(remote, desired);
}

File: src/grace_platform/vapi/lib/drift.py
"""Merge-based drift detection (doc 08 §8.1).

A naive ``local`` vs ``remote`` diff is permanently red: Vapi materialises every server
default and adds new ones over time. Instead we compare ``remote`` against
``deep_merge(remote, local)`` — so drift is non-empty **iff a key we actually declare has
a different value remotely**. Keys the server adds and we do not declare are invisible.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

VOLATILE = frozenset(
    {
        "id",
        "orgId",
        "createdAt",
        "updatedAt",
        "isServerUrlSecretSet",
        "credentialId",  # env-injected, instance-specific — masked, not compared
    }
)

SORT_ARRAYS_AT = frozenset({"toolIds", "structuredOutputIds", "serverMessages"})

FORBIDDEN_DRIFT: tuple[str, ...] = (
    "firstMessage",
    "serverMessages",
    "server.url",
    "model.messages.0.content",
    "compliancePlan.hipaaEnabled",
    "compliancePlan.pciEnabled",
    "artifactPlan.transcriptPlan.enabled",
)
"""Any drift here is a hard failure, never a warning: these carry compliance or routing
meaning (doc 08 §8.1)."""


def deep_merge(remote: Any, local: Any) -> Any:
    """``local`` overlays ``remote``. Arrays are replaced wholesale, never merged."""
    if not isinstance(remote, dict) or not isinstance(local, dict):
        return local
    out = dict(remote)
    for k, v in local.items():
        prev = out.get(k)
        out[k] = deep_merge(prev, v) if isinstance(prev, dict) and isinstance(v, dict) else v
    return out


def normalise(value: Any, key: str = "") -> Any:
    """Strips volatile keys, sorts unordered arrays, canonicalises for comparison."""
    if isinstance(value, list):
        items = [normalise(v) for v in value]
        if key in SORT_ARRAYS_AT:
            return sorted(items, key=lambda x: json.dumps(x, sort_keys=True))
        return items
    if isinstance(value, dict):
        out: dict[str, Any] = {}
        for k in sorted(value):
            if k in VOLATILE:
                continue
            v = value[k]
            if v is None:  # null ≡ absent
                continue
            out[k] = normalise(v, k)
        return out
    if isinstance(value, str):
        return value.rstrip()
    # JSON has one number type; Python has two. `1.0` from a config file and `1` echoed
    # back by Vapi are the same value, but compare unequal — which made the drift check
    # permanently red on the very first Python run. Collapse integral floats to int.
    if isinstance(value, float) and value.is_integer():
        return int(value)
    return value


@dataclass(frozen=True, slots=True)
class DriftEntry:
    path: str
    remote: Any
    desired: Any
    forbidden: bool


def _is_forbidden(path: str) -> bool:
    return any(path == f or path.startswith(f + ".") for f in FORBIDDEN_DRIFT)


def diff(remote: Any, desired: Any, base: str = "") -> list[DriftEntry]:
    """Reports only paths present in ``desired`` — i.e. paths we declare."""
    out: list[DriftEntry] = []

    if isinstance(remote, dict) and isinstance(desired, dict):
        for k in desired:
            out.extend(diff(remote.get(k), desired.get(k), f"{base}.{k}" if base else k))
        return out

    if json.dumps(remote, sort_keys=True) != json.dumps(desired, sort_keys=True):
        out.append(DriftEntry(base, remote, desired, _is_forbidden(base)))
    return out


def compute_drift(remote_raw: Any, local_raw: Any) -> list[DriftEntry]:
    return diff(normalise(remote_raw), normalise(deep_merge(remote_raw, local_raw)))
